### ISSUE TITLE
Moving to userspace (originally: bringing kernel drivers upstream)

### ISSUE THREAD

**Issue by Lyude** [openrazer/openrazer#311]
Hi! I am a developer from Red Hat who primarily works on driver support for Linux. I recently actually purchased a Razer Blade Stealth for myself and stumbled upon your project. It's really nice to see that there's been a good bit of work reverse engineering these devices done here. However, it would be really nice to eventually see these machines actually supported out of the box on Linux distributions, and doing that would require that you guys eventually merge your drivers upstream with the kernel itself. Do you guys happen to have any plans for doing that in the future?

**Comment by z3ntu** [comment]
You are probably interested in #29

**Comment by terricain** [comment]
Hey, Yes and No, #29 is correct to some degree. 

The majority of the keyboard driver part can be done in userspace as the endpoint that deals with lighting does not transmit key events.

From what I understand  you'll need to unbind a USB endpoint from whatever kernel driver is using it before you can interact with it from userspace. That wont work for cases like the mice which use the same endpoint for transmitting movement data as well as receiving packets to control lighting.

It would be nice to get what we have into the kernel but it would require some big changes which is not too bad, its mostly once its in the kernel its going to become increasingly difficult to modify, which in its self isn't a bad thing either.

Currently the aim is just to support as many devices as possible once that's slowed down a bit I don't see any issue with it going into the kernel. Were using dkms at the moment to deal with all the kernel stuff which seems to be doing its job for the most part (except when secure boot makes the odd appearance, which would be solved if it were in-tree).

**Comment by whompy** [comment]
I'm with lyude on this one. I'd love it in the kernel as it's the only dkms module on my system. Having the kernel driver in tree would be most excellent.

That said, I understand that it's best to get something in a good, solid state before stabilizing the api in upstream. As a follower of the project, I'd be interested to know if there are specific milestones you'd like before upstreaming. 

Again, I understand the logic and what you're doing, just curious if there's a long-term plan.

**Comment by ukyg9e5r6k7gubiekd6** [comment]
I too am interested in fixing the kernel-mode parts and upstreaming them (see #428, which partly duplicates this).
I'm intending to just scratch my itch, support for control over the LEDs on Razer mice (I don't have any other Razer devices) shortly, when I make the time. Especially as I understand from terrycain's explanation that that has to be done in-kernel, due to multiplexing lighting and movement data on the same endpoints.

**Comment by terricain** [comment]
> due to multiplexing lighting and movement data on the same endpoints.

Not exactly, take a look at the `hid` branch were slowly moving stuff to do as much of the work in userspace as possible. For the mice and keyboards, all that will end up in the kernel I hope is some messing around/fixing of key/button events

**Comment by knghtbrd** [comment]
This issue's been dormant for about three years now, and it doesn't look like the hid branch has seen any work to pull in changes from master in about that long. Is the notion of pushing as much as possible into userspace and the approach taken by the hid branch being more or less abandoned?

What does this mean for these drivers becoming mainstream? "Tainted kernels", etc.

**Comment by z3ntu** [comment]
I would love for most of the functionality to be implemented in user space (see my experimental project https://github.com/z3ntu/razer_test) but it's not really possible to change the implementation for 100+ already supported devices and keep them working without any regressions.

**Comment by FFY00** [comment]
If you know how the devices behave, you can emulate the devices using UHID, run the tests against the current driver and make sure everything is working as expected, and then you can port the whole thing to userspace and make sure everything still works as expected. You'd still want to have some people testing this on real hardware at some point, but this approach would allow you to port the drivers without having the devices physically.

**Comment by z3ntu** [comment]
If you could develop this tool then that would be great, but I don't have too much time / motivation to work on such a thing (at the moment).

**Comment by FFY00** [comment]
We have a PoC at https://github.com/libratbag/ratbag-emu, but I will be redesigning everything from scratch on top of another backend library. You can make use of that work later if you want. I don't have any razer devices, though I'd be interested in bringing support to libratbag, and I have limited time so I probably won't be contributing to this upstream.

**Comment by Lyude** [comment]
Hi! Uh, I'm back haha. It's been a number of years and I've learned quite a lot about, well, everything kernel/linux related. I'm back in the situation of trying to get myself a new mouse and unfortunately my beloved G-Skill MX780 is no longer being made, which means there's literally no ergonomic left handed mouse on the market that meets my needs and I have now arrived back at considering razer. In the interim between when I last posted and now, I actually wrote part of the libratbag driver for the g-skill mouse so I've learned a bit about gaming mice in particular.

Taking a look at this thread, I see that one of the things mentioned here is regressions: this is rather interesting to me because with ratbag, what I saw was that pretty much every single gaming mouse uses some variant of HID in order for actually controlling lights/buttons on the mouse. Which makes me curious what the actual difference is between that and Razer mice? If I'm understanding correctly from:

https://github.com/openrazer/openrazer/blob/master/driver/razercommon.c#L20

It looks like all you guys are doing is sending `HID_REQ_GET_REPORT`/`HID_REQ_SET_REPORT`, except from the kernel instead of through the userspace interfaces used for doing this. Which also is to say: if you are able to confirm with even just one mouse that things work with the new driver, I don't actually see that much likelihood for a regression unless there's something else sneaky going on here I didn't notice. I don't believe anything should actually change from the perspective of the mouse (/whatever razer device we're controlling) either. The only gotcha I can think of is just making sure you include the right udev rules so that you can access the proper device nodes from userspace.

If there was a chance for a regression as well, it sounds like a good path forward would be to abstract the helpers you're using for sending/receiving data from the mouse so they can go through either interface. Regardless though, is this something the project would be interested in getting a contribution for? Regarding the UHID stuff as well, if you're interested in testing replays of devices it's actually not terribly difficult to implement that sort of thing. I did this ages ago by writing a program for recording/replaying PS/2 devices over serio, and I just had the userspace program play back the data that was sent by the device (along with reproducing any delays that occurred during the recording, e.g. time spent sleeping while the host was waiting for data from the device).

**Comment by FFY00** [comment]
FWIW, I have the same opinion.

**Comment by z3ntu** [comment]
@Lyude If you have a workable plan on how to actually do the change, then I'm happy to try and give a clear answer from my maintainer perspective.

Yes, it should be even possible to take most parts of the driver parts and bend them into working with hidapi/libusb from user space. There's existing projects doing something similar, taking OpenRazer code and making it work with macOS (from user space afaik) so it's definitely not impossible. But I myself own a very limited number of devices, so getting a good sample size and the different ways some devices seem to behave (see e.g. `expected_subclass`, `expected_protocol` and stuff like that) is difficult.

But yes, I'm 100% interested in moving most parts of the driver to user space (input handling will need to remain in the kernel I think). In 2018/2019 I was myself working a bit on a separate replacement: https://github.com/z3ntu/razer_test

**Comment by Lyude** [comment]
> [@Lyude](https://github.com/Lyude) If you have a workable plan on how to actually do the change, then I'm happy to try and give a clear answer from my maintainer perspective.
> 
> Yes, it should be even possible to take most parts of the driver parts and bend them into working with hidapi/libusb from user space. There's existing projects doing something similar, taking OpenRazer code and making it work with macOS (from user space afaik) so it's definitely not impossible. But I myself own a very limited number of devices, so getting a good sample size and the different ways some devices seem to behave (see e.g. `expected_subclass`, `expected_protocol` and stuff like that) is difficult.
> 
> But yes, I'm 100% interested in moving most parts of the driver to user space (input handling will need to remain in the kernel I think). In 2018/2019 I was myself working a bit on a separate replacement: https://github.com/z3ntu/razer_test

No promise on deadlines but I'll look into getting myself a left handed razer mouse then :)

**Comment by Lyude** [comment]
(I got the mouse! So this is officially on my to-do list when I've got the time!)

**Comment by Lyude** [comment]
JFYI - what sort of proof for avoiding regressions would actually be needed here, btw? I'm not terribly confident that we could actually go through every device one-by-one to confirm there are no regressions, but I also don't really think it would be necessary if we can just get enough devices to cover the majority of the functions that the driver already supports and then just compare the actual HID traces going across the line between the in-kernel version of openrazer and the (future) userspace version of openrazer. I think this is true as well because `/dev/hidraw*` is really just a character-device front-end to the same kernel interface that the kernel module is already using to communicate with mice/keyboards/anything over hid.

If that's something we'd be nervous about doing, would anyone be against me seeing if I could just make it configurable whether it uses the kernel module or not?
